### DISCUSSION TITLE
Ways to prevent 100% cloning when onboarding

### DISCUSSION THREAD

**Discussion by RahulGautamSingh** [renovatebot/renovate#17026] [Suggest an Idea]
Currently we perform cloning each onboarding run for repos that aren't onboarded. The cloning occurs because of the two checks `isOnboarded` and the if-block after `onboardingPrExists` + the `checkoutBranch`. We can start by refactoring the functions one by one:
1. isOnboarded 
- Current Flow: 
    1. check config options to detect if repo is already onboarded
    2. get configFileName from cache and use api-requests to check of file exists
    3. clone the repo and check for configFile (fns: configFileExists/packageJsonFileExists)
    4. search for closed PR and configOptions which makes configFile optional
    5. finally ensureComment notifying the user that they may have to add a configFile manually or rename the closed/mergedPR
  
 We need some checks before step-3. I came up with these 2 solutions:
    1. Check for an open onboardingPr and if it exists then return false, we can add a timeLimit to this so that we don't wait too much
    2. Cache the result of configFileExists/packageJsonFileExists function reuse it till the parentBranchSha doesn't change.

**Comment by rarkins** [comment]
If the default branch SHA has changed then I think we always need to clone, in case dependencies have changed.

But if it's unchanged then any cached isOnboarded we store should still be valid.

The next question is: has the user made any changes to the onboarding branch, e.g. added additional config. In such a case we need to clone also.

But if there's no change to the default or onboarding branch SHAs, I'm not sure there's any other conditions where we need to clone?

**Reply by RahulGautamSingh** [reply]
Yes we don't need to worry for any other condition if we check both the SHAs. 

Regarding changes to onboardingBranch we itthem using isBranchModified call which in this case can't use cached result because we don't cache onboardingBranch details. I think we can either store the onboardingBranch details like we do for update-branches till the `not-onboarded` and delete once `onboarded`.

**Reply by rarkins** [reply]
For onboarding repos I think we need to store in cache:
- Default branch SHA (should already be there)
- isOnboarded
- onboarding branch SHA (btw if this is set then we don't necessarily need isOnboarded boolean)
- extracted dependencies

If either SHA has changed then it's OK to do a clone like we do currently. But if not then we should avoid it and instead:
- Do lookups using cached extracted dependencies
- Check if PR needs updating

**Reply by rarkins** [reply]
BTW #17020 should speed up onboarding repos too

**Reply by RahulGautamSingh** [reply]
Will convert this to an issue, I think it should be done in 2 parts 
- one handles the introduction of onboardingBranch in repoCache + we use it to prevent cloning.
- another one handles the lookup part.
